// @gooddata/sdk-ui-ext/esm/internal/utils/dimensions.js
import { __spreadArrays } from "tslib";
// (C) 2007-2022 GoodData Corporation
import compact from "lodash/compact";
import { bucketAttributes, bucketIsEmpty, insightAttributes, insightBucket, insightTotals, MeasureGroupIdentifier, newDimension, newTwoDimensional, } from "@gooddata/sdk-model";
import { BucketNames, VisualizationTypes } from "@gooddata/sdk-ui";
function safeBucketAttributes(insight, idOrFun) {
    var matchingBucket = insightBucket(insight, idOrFun);
    return matchingBucket ? bucketAttributes(matchingBucket) : [];
}
export function getPivotTableDimensions(insight) {
    var measures = insightBucket(insight, BucketNames.MEASURES);
    var measuresItemIdentifiers = measures && !bucketIsEmpty(measures) ? [MeasureGroupIdentifier] : [];
    var rowAttributes = safeBucketAttributes(insight, BucketNames.ATTRIBUTE);
    var columnAttributes = safeBucketAttributes(insight, BucketNames.COLUMNS);
    return newTwoDimensional(__spreadArrays(rowAttributes, insightTotals(insight)), __spreadArrays(columnAttributes, measuresItemIdentifiers));
}
function getPieDonutFunnelDimensions(insight) {
    var viewByAttributes = safeBucketAttributes(insight, BucketNames.VIEW);
    return viewByAttributes.length
        ? newTwoDimensional([MeasureGroupIdentifier], viewByAttributes)
        : newTwoDimensional([], [MeasureGroupIdentifier]);
}
function getBarColumnDimensions(insight) {
    var viewByAttributes = safeBucketAttributes(insight, BucketNames.VIEW);
    var stackByAttributes = safeBucketAttributes(insight, BucketNames.STACK);
    return stackByAttributes.length
        ? newTwoDimensional(stackByAttributes, __spreadArrays(viewByAttributes, [MeasureGroupIdentifier]))
        : newTwoDimensional([MeasureGroupIdentifier], viewByAttributes);
}
function getAreaDimensions(insight) {
    var _a;
    var viewByAttributes = safeBucketAttributes(insight, BucketNames.VIEW);
    var stackByAttributes = safeBucketAttributes(insight, BucketNames.STACK);
    var seriesAttribute = viewByAttributes[0];
    var sliceAttribute = (_a = stackByAttributes[0]) !== null && _a !== void 0 ? _a : viewByAttributes[1];
    return sliceAttribute
        ? newTwoDimensional([sliceAttribute], compact([seriesAttribute, MeasureGroupIdentifier]))
        : newTwoDimensional([MeasureGroupIdentifier], compact([seriesAttribute]));
}
function getLineDimensions(insight) {
    var trendAttributes = safeBucketAttributes(insight, BucketNames.TREND);
    var segmentAttributes = safeBucketAttributes(insight, BucketNames.SEGMENT);
    return segmentAttributes.length
        ? newTwoDimensional(segmentAttributes, __spreadArrays(trendAttributes, [MeasureGroupIdentifier]))
        : newTwoDimensional([MeasureGroupIdentifier], trendAttributes);
}
export function getHeadlinesDimensions() {
    return [newDimension([MeasureGroupIdentifier])];
}
function getScatterDimensions(insight) {
    var attributes = safeBucketAttributes(insight, BucketNames.ATTRIBUTE);
    return newTwoDimensional(attributes, [MeasureGroupIdentifier]);
}
function getHeatmapDimensions(insight) {
    var viewByAttributes = safeBucketAttributes(insight, BucketNames.VIEW);
    var stackByAttributes = safeBucketAttributes(insight, BucketNames.STACK);
    return newTwoDimensional(viewByAttributes, __spreadArrays(stackByAttributes, [MeasureGroupIdentifier]));
}
function getBulletComboDimensions(insight) {
    var viewByAttributes = safeBucketAttributes(insight, BucketNames.VIEW);
    return newTwoDimensional([MeasureGroupIdentifier], viewByAttributes);
}
function getBubbleDimensions(insight) {
    var viewByAttributes = safeBucketAttributes(insight, BucketNames.VIEW);
    return newTwoDimensional(viewByAttributes, [MeasureGroupIdentifier]);
}
/**
 * Generates dimensions based on buckets and visualization objects.
 *
 * @param insight - insight being visualized
 * @param type - visualization type string
 * @internal
 */
export function generateDimensions(insight, type) {
    switch (type) {
        case VisualizationTypes.TABLE:
            return getPivotTableDimensions(insight);
        case VisualizationTypes.PIE:
        case VisualizationTypes.DONUT:
        case VisualizationTypes.FUNNEL:
            return getPieDonutFunnelDimensions(insight);
        case VisualizationTypes.TREEMAP:
            return getTreemapDimensions(insight);
        case VisualizationTypes.LINE:
            return getLineDimensions(insight);
        case VisualizationTypes.AREA:
            return getAreaDimensions(insight);
        case VisualizationTypes.BAR:
        case VisualizationTypes.COLUMN:
            return getBarColumnDimensions(insight);
        case VisualizationTypes.BULLET:
        case VisualizationTypes.COMBO:
        case VisualizationTypes.COMBO2:
            return getBulletComboDimensions(insight);
        case VisualizationTypes.HEADLINE:
            return getHeadlinesDimensions();
        case VisualizationTypes.SCATTER:
            return getScatterDimensions(insight);
        case VisualizationTypes.HEATMAP:
            return getHeatmapDimensions(insight);
        case VisualizationTypes.BUBBLE:
            return getBubbleDimensions(insight);
    }
    return [];
}
export function generateStackedDimensions(insight) {
    var viewByAttributes = safeBucketAttributes(insight, BucketNames.ATTRIBUTE);
    var stackByAttributes = safeBucketAttributes(insight, BucketNames.STACK);
    return newTwoDimensional(stackByAttributes, __spreadArrays(viewByAttributes, [MeasureGroupIdentifier]));
}
function getTreemapDimensions(insight) {
    var attributes = insightAttributes(insight);
    return attributes.length === 1
        ? newTwoDimensional([MeasureGroupIdentifier], attributes)
        : newTwoDimensional(attributes, [MeasureGroupIdentifier]);
}
