// @gooddata/sdk-ui-ext/esm/internal/utils/embeddingCodeGenerator/insightToPropsConverter/convertor.js
// (C) 2022 GoodData Corporation
import { insightBucket } from "@gooddata/sdk-model";
import toPairs from "lodash/toPairs";
/**
 * Conversion based on the whole {@link @gooddata/sdk-model#IInsightDefinition}.
 */
export function insightConversion(propName, propType, insightItemAccessor) {
    return {
        propName: propName,
        propType: propType,
        itemAccessor: insightItemAccessor,
    };
}
/**
 * Conversion based on a single bucket.
 */
export function bucketConversion(propName, propType, bucketName, bucketItemAccessor) {
    return {
        propName: propName,
        propType: propType,
        itemAccessor: function (insight) {
            var bucket = insightBucket(insight, bucketName);
            return bucket && bucketItemAccessor(bucket);
        },
    };
}
/**
 * Creates an InsightToProps converter.
 *
 * @remarks
 * This makes the conversion as declarative as possible avoiding any explicit logic in the call sites,
 * rather using specialized object to describe parts of the conversion.
 *
 * @param conversionSpec - Specification of the insight to props conversion
 * @returns function that can be used to convert a given insight to props for some visualization type
 */
export function getInsightToPropsConverter(conversionSpec) {
    return function (insight, ctx) {
        return toPairs(conversionSpec).reduce(function (acc, _a) {
            var propName = _a[0], conversion = _a[1];
            acc[propName] = {
                value: conversion.itemAccessor(insight, ctx),
                meta: conversion.propType,
            };
            return acc;
        }, {});
    };
}
