// @gooddata/sdk-ui-dashboard/esm/presentation/scheduledEmail/DefaultScheduledEmailManagementDialog/utils.js
// (C) 2022 GoodData Corporation
import { isDashboardAttachment, isWidgetAttachment, } from "@gooddata/sdk-backend-spi";
import { Icon } from "@gooddata/sdk-ui-kit";
import compact from "lodash/compact";
export var getRecipientsLabel = function (intl, recipients, currentUserEmail) {
    var numberOfRecipients = recipients.length;
    var recipientsLabel = numberOfRecipients === 1 && recipients[0] === currentUserEmail
        ? intl.formatMessage({ id: "dialogs.schedule.management.recipients.onlyYou" })
        : intl.formatMessage({
            id: "dialogs.schedule.management.recipients",
        }, { n: numberOfRecipients });
    return recipientsLabel;
};
export var getAttachmentType = function (intl, attachments) {
    var hasDashboardAttachment = attachments.some(isDashboardAttachment);
    var numberOfWidgetAttachments = attachments.filter(isWidgetAttachment).length;
    if (hasDashboardAttachment && numberOfWidgetAttachments) {
        return {
            AttachmentIcon: Icon.Many,
            attachmentLabel: intl.formatMessage({ id: "dialogs.schedule.management.attachments.mixed" }, { n: numberOfWidgetAttachments }),
        };
    }
    else if (numberOfWidgetAttachments) {
        return {
            AttachmentIcon: Icon.Widget,
            attachmentLabel: intl.formatMessage({ id: "dialogs.schedule.management.attachments.widgets" }, { n: numberOfWidgetAttachments }),
        };
    }
    else {
        return {
            AttachmentIcon: Icon.Dashboard,
            attachmentLabel: intl.formatMessage({ id: "dialogs.schedule.management.attachments.dashboard" }),
        };
    }
};
export var getFormatsLabel = function (attachments) {
    var hasPdfFormat = attachments.some(isDashboardAttachment);
    var widgetAttachments = attachments.filter(isWidgetAttachment);
    var hasCsvFormat = widgetAttachments.some(function (attachment) { return attachment.formats.includes("csv"); });
    var hasXlsxFormat = widgetAttachments.some(function (attachment) { return attachment.formats.includes("xlsx"); });
    return compact([hasPdfFormat && "PDF", hasCsvFormat && "CSV", hasXlsxFormat && "XLSX"]).join(", ");
};
