// @gooddata/api-client-tiger/esm/client.js
// (C) 2019-2022 GoodData Corporation
import { tigerExecutionClientFactory } from "./execution";
import { tigerExecutionResultClientFactory } from "./executionResult";
import { LabelElementsConfiguration, LabelElementsBaseApi, tigerLabelElementsClientFactory, } from "./labelElements";
import { tigerValidObjectsClientFactory } from "./validObjects";
import { setAxiosAuthorizationToken } from "./axios";
import { tigerLayoutClientFactory } from "./layout";
import { tigerAfmExplainClientFactory } from "./explain";
import { MetadataConfiguration, MetadataBaseApi, tigerEntitiesObjectsClientFactory, } from "./entitiesObjects";
export { tigerExecutionClientFactory, tigerEntitiesObjectsClientFactory, tigerExecutionResultClientFactory, tigerLabelElementsClientFactory, tigerValidObjectsClientFactory, tigerLayoutClientFactory, tigerAfmExplainClientFactory, MetadataConfiguration, MetadataBaseApi, LabelElementsConfiguration, LabelElementsBaseApi, };
/**
 * Tiger execution client
 *
 */
export var tigerClientFactory = function (axios) {
    var execution = tigerExecutionClientFactory(axios);
    var executionResult = tigerExecutionResultClientFactory(axios);
    var labelElements = tigerLabelElementsClientFactory(axios);
    var validObjects = tigerValidObjectsClientFactory(axios);
    var declarativeLayout = tigerLayoutClientFactory(axios);
    var explain = tigerAfmExplainClientFactory(axios);
    var entities = tigerEntitiesObjectsClientFactory(axios);
    return {
        axios: axios,
        execution: execution,
        executionResult: executionResult,
        labelElements: labelElements,
        validObjects: validObjects,
        declarativeLayout: declarativeLayout,
        explain: explain,
        entities: entities,
        setApiToken: function (token) {
            setAxiosAuthorizationToken(axios, token);
        },
    };
};
