// @gooddata/sdk-backend-spi/esm/workspace/dateFilterConfigs/types.js
// (C) 2007-2022 GoodData Corporation
import isEmpty from "lodash/isEmpty";
var dateFilterGranularity = [
    "GDC.time.date",
    "GDC.time.week_us",
    "GDC.time.month",
    "GDC.time.quarter",
    "GDC.time.year",
];
/**
 * Type-guard testing whether the provided object is an instance of {@link DateFilterGranularity}.
 * @public
 */
export var isDateFilterGranularity = function (obj) {
    return !isEmpty(obj) && dateFilterGranularity.some(function (granularity) { return granularity === obj; });
};
/**
 * Type-guard testing whether the provided object is an instance of {@link IAllTimeDateFilterOption}.
 * @public
 */
export var isAllTimeDateFilterOption = function (obj) {
    return !isEmpty(obj) && obj.type === "allTime";
};
/**
 * Type-guard testing whether the provided object is an instance of {@link IAbsoluteDateFilterForm}.
 * @public
 */
export var isAbsoluteDateFilterForm = function (obj) {
    return !isEmpty(obj) && obj.type === "absoluteForm";
};
/**
 * Type-guard testing whether the provided object is an instance of {@link IAbsoluteDateFilterPreset}.
 * @public
 */
export var isAbsoluteDateFilterPreset = function (obj) {
    return !isEmpty(obj) && obj.type === "absolutePreset";
};
/**
 * Type-guard testing whether the provided object is an instance of {@link IRelativeDateFilterForm}.
 * @public
 */
export var isRelativeDateFilterForm = function (obj) {
    return !isEmpty(obj) && obj.type === "relativeForm";
};
/**
 * Type-guard testing whether the provided object is an instance of {@link IRelativeDateFilterPreset}.
 * @public
 */
export var isRelativeDateFilterPreset = function (obj) {
    return !isEmpty(obj) && obj.type === "relativePreset";
};
