// @gooddata/sdk-ui-ext/esm/internal/components/pluggableVisualizations/bulletChart/BulletChartDescriptor.js
import { __extends } from "tslib";
import { BucketNames, getIntersectionPartAfter } from "@gooddata/sdk-ui";
import { PluggableBulletChart } from "./PluggableBulletChart";
import { BaseChartDescriptor } from "../baseChart/BaseChartDescriptor";
import { modifyBucketsAttributesForDrillDown, addIntersectionFiltersToInsight } from "../drillDownUtil";
import { drillDownFromAttributeLocalId } from "../../../utils/ImplicitDrillDownHelper";
import { filtersInsightConversion, getInsightToPropsConverter, getReactEmbeddingCodeGenerator, multipleAttributesBucketConversion, singleAttributeOrMeasureBucketConversion, sortsInsightConversion, } from "../../../utils/embeddingCodeGenerator";
import { chartAdditionalFactories, chartConfigInsightConversion } from "../chartCodeGenUtils";
var BulletChartDescriptor = /** @class */ (function (_super) {
    __extends(BulletChartDescriptor, _super);
    function BulletChartDescriptor() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.getEmbeddingCode = getReactEmbeddingCodeGenerator({
            component: {
                importType: "named",
                name: "BulletChart",
                package: "@gooddata/sdk-ui-charts",
            },
            insightToProps: getInsightToPropsConverter({
                primaryMeasure: singleAttributeOrMeasureBucketConversion("primaryMeasure", BucketNames.MEASURES),
                targetMeasure: singleAttributeOrMeasureBucketConversion("targetMeasure", BucketNames.SECONDARY_MEASURES),
                comparativeMeasure: singleAttributeOrMeasureBucketConversion("comparativeMeasure", BucketNames.TERTIARY_MEASURES),
                viewBy: multipleAttributesBucketConversion("viewBy", BucketNames.VIEW),
                filters: filtersInsightConversion("filters"),
                sortBy: sortsInsightConversion("sortBy"),
                config: chartConfigInsightConversion("config"),
            }),
            additionalFactories: chartAdditionalFactories,
        });
        return _this;
    }
    BulletChartDescriptor.prototype.getFactory = function () {
        return function (params) { return new PluggableBulletChart(params); };
    };
    BulletChartDescriptor.prototype.applyDrillDown = function (insight, drillDownContext) {
        var withFilters = this.addFiltersForBullet(insight, drillDownContext.drillDefinition, drillDownContext.event);
        return modifyBucketsAttributesForDrillDown(withFilters, drillDownContext.drillDefinition);
    };
    BulletChartDescriptor.prototype.addFiltersForBullet = function (insight, drillConfig, event) {
        var clicked = drillDownFromAttributeLocalId(drillConfig);
        var cutIntersection = getIntersectionPartAfter(event.drillContext.intersection, clicked);
        return addIntersectionFiltersToInsight(insight, cutIntersection);
    };
    return BulletChartDescriptor;
}(BaseChartDescriptor));
export { BulletChartDescriptor };
