// @gooddata/sdk-ui-ext/esm/internal/components/pluggableVisualizations/geoChart/GeoPushpinChartDescriptor.js
import { __extends } from "tslib";
import { BucketNames } from "@gooddata/sdk-ui";
import { PluggableGeoPushpinChart } from "./PluggableGeoPushpinChart";
import { DASHBOARD_LAYOUT_DEFAULT_VIS_HEIGHT, MIDDLE_VISUALIZATION_HEIGHT } from "../constants";
import { BaseChartDescriptor } from "../baseChart/BaseChartDescriptor";
import { filtersInsightConversion, getInsightToPropsConverter, getReactEmbeddingCodeGenerator, insightConversion, singleAttributeBucketConversion, singleAttributeOrMeasureBucketConversion, sortsInsightConversion, } from "../../../utils/embeddingCodeGenerator";
import { geoConfigFromInsight } from "./geoConfigFromInsight";
import { chartAdditionalFactories } from "../chartCodeGenUtils";
var GeoPushpinChartDescriptor = /** @class */ (function (_super) {
    __extends(GeoPushpinChartDescriptor, _super);
    function GeoPushpinChartDescriptor() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.getEmbeddingCode = getReactEmbeddingCodeGenerator({
            component: {
                importType: "named",
                name: "GeoPushpinChart",
                package: "@gooddata/sdk-ui-geo",
            },
            insightToProps: getInsightToPropsConverter({
                location: singleAttributeBucketConversion("location", BucketNames.LOCATION),
                size: singleAttributeOrMeasureBucketConversion("size", BucketNames.SIZE),
                color: singleAttributeOrMeasureBucketConversion("color", BucketNames.COLOR),
                segmentBy: singleAttributeBucketConversion("segmentBy", BucketNames.SEGMENT),
                filters: filtersInsightConversion("filters"),
                sortBy: sortsInsightConversion("sortBy"),
                config: insightConversion("config", {
                    typeImport: {
                        importType: "named",
                        name: "IGeoConfig",
                        package: "@gooddata/sdk-ui-geo",
                    },
                    cardinality: "scalar",
                }, geoConfigFromInsight),
            }),
            additionalFactories: chartAdditionalFactories,
        });
        return _this;
    }
    GeoPushpinChartDescriptor.prototype.getFactory = function () {
        return function (params) { return new PluggableGeoPushpinChart(params); };
    };
    GeoPushpinChartDescriptor.prototype.getSizeInfo = function (_insight, layoutDescriptor, settings) {
        return {
            width: {
                default: 6,
                min: 6,
                max: layoutDescriptor.gridColumnsCount,
            },
            height: {
                default: this.getDefaultHeight(settings.enableKDWidgetCustomHeight),
                min: this.getDefaultHeight(settings.enableKDWidgetCustomHeight),
                max: this.getMaxHeight(settings.enableKDWidgetCustomHeight),
            },
        };
    };
    GeoPushpinChartDescriptor.prototype.getMinHeight = function (enableCustomHeight) {
        if (!enableCustomHeight) {
            return DASHBOARD_LAYOUT_DEFAULT_VIS_HEIGHT;
        }
        return MIDDLE_VISUALIZATION_HEIGHT;
    };
    return GeoPushpinChartDescriptor;
}(BaseChartDescriptor));
export { GeoPushpinChartDescriptor };
